eval/src/vespa/eval/instruction/dense_fast_rename_optimizer.cpp

```diff
@@ -10,25 +10,52 @@ using namespace tensor_function;
 
 namespace {
 
-bool is_dense_stable_rename(const ValueType &from_type, const ValueType &to_type,
-                            const std::vector<vespalib::string> &from,
-                            const std::vector<vespalib::string> &to)
+bool is_ascending(const std::vector<size_t> &values) {
+    for (size_t i = 1; i < values.size(); ++i) {
+        if (values[i-1] >= values[i]) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool is_stable_rename(const ValueType &from_type, const ValueType &to_type,
+                      const std::vector<vespalib::string> &from,
+                      const std::vector<vespalib::string> &to)
 {
-    if (!from_type.is_dense() ||
-        !to_type.is_dense() ||
-        (from.size() != to.size()))
-    {
+    if (from.size() != to.size()) {
         return false;
     }
     size_t npos = ValueType::Dimension::npos;
+    std::map<vespalib::string, size_t> name_to_new_idx;
```

Comment:
consider mapping old index to new index directly, also consider using separate maps for mapped and indexed dimensions and traversing the maps directly to verify stable orderings.